flytekit/common/tasks/sagemaker/custom_training_job_task.py

```diff
@@ -77,6 +88,69 @@ def __init__(
+        :param flytekit.models.literals.LiteralMap inputs:
+        :rtype: dict[Text, flytekit.models.common.FlyteIdlEntity]
+        :returns: This function must return a dictionary mapping 'filenames' to Flyte Interface Entities.  These
+            entities will be used by the engine to pass data from node to node, populate metadata, etc. etc..  Each
+            engine will have different behavior.  For instance, the Flyte engine will upload the entities to a remote
+            working directory (with the names provided), which will in turn allow Flyte Propeller to push along the
+            workflow.  Where as local engine will merely feed the outputs directly into the next node.
+        """
+        ret = super().execute(context, inputs)
+
+        if (
+            self._is_distributed()
+            and self._output_persist_predicate
+            and context.distributed_training_context
+            and self.output_persist_predicate(context.distributed_training_context) is True
+        ):
+            return ret
+        else:
+            return {}
+
+    def _execute_user_code(self, context, inputs):
+        """
+        :param flytekit.engines.common.EngineContext context:
+        :param dict[Text, T] inputs: This variable is a bit of a misnomer, since it's both inputs and outputs. The
+            dictionary passed here will be passed to the user-defined function, and will have values that are a
+            variety of types.  The T's here are Python std values for inputs.  If there isn't a native Python type for
+            something (like Schema or Blob), they are the Flyte classes.  For outputs they are OutputReferences.
+            (Note that these are not the same OutputReferences as in BindingData's)
+        :rtype: Any: the returned object from user code.
+        :returns: This function must return a dictionary mapping 'filenames' to Flyte Interface Entities.  These
+            entities will be used by the engine to pass data from node to node, populate metadata, etc. etc..  Each
+            engine will have different behavior.  For instance, the Flyte engine will upload the entities to a remote
+            working directory (with the names provided), which will in turn allow Flyte Propeller to push along the
+            workflow.  Where as local engine will merely feed the outputs directly into the next node.
+        """
+
+        return _exception_scopes.user_entry_point(self.task_function)(
+            _ExecutionParameters(
+                execution_date=context.execution_date,
+                # TODO: it might be better to consider passing the full struct
+                execution_id=_six.text_type(WorkflowExecutionIdentifier.promote_from_model(context.execution_id)),
+                stats=context.stats,
+                logging=context.logging,
+                tmp_dir=context.working_directory,
+                distributed_training_context=context.distributed_training_context,
```

Comment:
I think this is where you should actually attempt to build the context instead?

Reply:
Unfortunately, the distributed_training_context is needed not only by user task_function, but also the execute() function of the CustomTrainingJobTask. If we create `distributed_training_context` here, the execute() function would not be able to access it. We can, of course, create it from the env var again, but I am not sure if that is preferred.

Reply:
What I can do is to fill `distributed_training_context` into the context object only when CustomTrainingJobTasks' execute() is called. But again I am not sure if that is preferred

Reply:
Update: @EngHabu and I discussed and agreed to build a training-job-specific context here: https://github.com/lyft/flytekit/pull/173/files#diff-3d88a183363264f577728ca8af28fd16R121